tests/Unit/Utilities/CssDocumentTest.php

```diff
@@ -0,0 +1,403 @@
+            'with space after' => ['p '],
+            'with line feed after' => ["p\n"],
+            'with Windows line ending after' => ["p\r\n"],
+            'with TAB after' => ["p\t"],
+            'minified' => ['p>.classy'],
+            'with line feeds within' => ["p\n>\n.classy"],
+        ];
+    }
+
+    /**
+     * @test
+     *
+     * @param string $declarations
+     *
+     * @dataProvider provideDeclarations
+     */
+    public function parsesDeclarations(string $declarations): void
+    {
+        $css = 'p {' . $declarations . '}';
+        $subject = new CssDocument($css);
+
+        $result = $subject->getStyleRulesData([]);
+
+        self::assertCount(1, $result);
+        $trimmedResultantDeclarations = \trim($result[0]['declarations']);
+        self::assertSame(\trim($declarations), $trimmedResultantDeclarations);
+    }
+
+    /**
+     * @return array<string, array<int, string>>
+     */
+    public function provideDeclarations(): array
+    {
+        return [
+            'colour (named)' => ['color: green;'],
+            'colour (RGB hex)' => ['color: #0F0;'],
+            'colour (RGBA)' => ['color: rgba(51, 170, 51, .1);'],
+            'length (px)' => ['margin-top: 10px;'],
+            'length (%)' => ['margin-right: 10%;'],
+            'shorthand' => ['margin: 0 auto;'],
+            'shorthand with mixed types' => ['background: no-repeat center/80% url("../img/image.png");'],
+            'calc' => ['margin-right: calc(10% + 2px);'],
+            'vendor property' => ['-webkit-appearance: button;'],
+            'vendor value' => [
+                'background-image: '
+                . '-webkit-gradient(linear, left top, left bottom, color-stop(0.05, #83bc38), color-stop(1, #57873b));',
+            ],
+            'minified' => ['color:green'],
+            'with space before' => [' color: green;'],
+            'with line feed before' => ["\ncolor: green;"],
+            'with Windows line ending before' => ["\r\ncolor: green;"],
+            'with TAB before' => ["\tcolor: green;"],
+            'with line feed within' => ["color:\ngreen;"],
+            'with line feed after' => ["color: green;\n"],
+            'multiple' => ['color: green; text-decoration: underline;'],
+            'multiple (minified)' => ['color:green;text-decoration:underline'],
+            'multiple (separated by line feed)' => ["color: green;\ntext-decoration: underline;"],
+            'multiple (separated by Windows line ending)' => ["color: green;\r\ntext-decoration: underline;"],
+            'multiple (separated by TAB)' => ["color: green;\ttext-decoration: underline;"],
+        ];
+    }
+
+    /**
+     * @test
+     *
+     * @param string $mediaQuery
+     * @param string $optionalSeparator
+     * @param string $separator
+     *
+     * @dataProvider provideMediaQueryAndSeparators
+     */
+    public function parsesMediaRule(string $mediaQuery, string $optionalSeparator = '', string $separator = ' '): void
```

Comment:
I propose we split this into two tests and two data providers: one for different types, and one for the formatting issues. Then we don't need any parameters to be optional anymore.

Reply:
Good idea.  Turns out the test with whitespace parameters doesn't need the media query parameter either.